### PR TITLE
feat: Poseidon2 implementation and availabe as Merkle Hash

### PR DESCRIPTION
**Pull request by alxiong**
closes: #704 

When we merge, I'd like to merge with commits (leave intermediate steps in the git history) if possible, as I try to make each commit self-contained in scope. 

### This PR: 

- Add `struct Poseidon2` and its permutation `::permute()` and `permute_mut()`
- Add `trait Poseidon2Params` whose generic parameter is much cleaner than plonky3, while sufficient for our purposes
- Add concrete parameters (2 instances for Bls12-381, 1 instance for Bn254) in `poseidon2/constants/` and add test vectors (generated by reference impl by `zkhash`)
- Implement `HasherAlgorithm` for `Poseidon2` so that it can be used in Merkle tree as the hash function candidate

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (main)
- [x] Linked to GitHub issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests
- [x] Updated relevant documentation in the code
- [x] Added relevant changelog entries to the `CHANGELOG.md` of touched crates.
- [x] Re-reviewed `Files changed` in the GitHub PR explorer

### CODE REVIEW

File: poseidon2/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,126 @@
+// Copyright (c) 2024 Espresso Systems (espressosys.com)
+// This file is part of the Jellyfish library.
+
+// You should have received a copy of the MIT License
+// along with the Jellyfish library. If not, see <https://mit-license.org/>.
+
+//! The Poseidon2 permutation.
```
Comment (mrain) [inline poseidon2/src/lib.rs:7]: Could we document here a list of implemented `(field, state_size)` tuple.

File: merkle_tree/src/prelude.rs [hunk 2]
```diff
@@ -52,6 +54,36 @@ pub type RescueLightWeightMerkleTree<F> = LightWeightMerkleTree<F, RescueHash<F>
 /// Example instantiation of a SparseMerkleTree indexed by I
 pub type RescueSparseMerkleTree<I, F> = UniversalMerkleTree<F, RescueHash<F>, I, 3, F>;
 
+// TODO: (alex) move this compression to CRHF and wrap with better API?
```
Comment (mrain) [inline merkle_tree/src/prelude.rs:57]: Agree that we should have a CRHF impl (and maybe sponge as well?

**Review by mrain — APPROVED** [review]
Generally LGTM with some comments.
- Should we also have a native benchmark?
- Circuit impl could be in a separate PR.

File: merkle_tree/src/prelude.rs (hunk 2 above)
Comment (alxiong) [reply merkle_tree/src/prelude.rs:57]: Yeah, let's add in a separate PR. This is already getting big

**Comment by alxiong** [comment]
## Benchmark

```
Poseidon2 over (Bls12_381::Fr, t=2)/1k iter
                        time:   [12.217 ms 12.254 ms 12.324 ms]
                        change: [-0.4638% +0.0421% +0.7025%] (p = 0.69 > 0.05)
                        No change in performance detected.
Found 1 outliers among 10 measurements (10.00%)
  1 (10.00%) high severe

Poseidon2 over (Bls12_381::Fr, t=2)/100k iter
                        time:   [1.2291 s 1.2487 s 1.2811 s]
                        change: [-0.1380% +1.5566% +4.0473%] (p = 0.11 > 0.05)
                        No change in performance detected.
Found 1 outliers among 10 measurements (10.00%)
  1 (10.00%) high severe

Poseidon2 over (Bls12_381::Fr, t=3)/1k iter
                        time:   [14.579 ms 14.650 ms 14.713 ms]
                        change: [-0.2505% +0.2707% +0.8794%] (p = 0.22 > 0.05)
                        No change in performance detected.
Found 1 outliers among 10 measurements (10.00%)
  1 (10.00%) high mild

Poseidon2 over (Bls12_381::Fr, t=3)/100k iter
                        time:   [1.4549 s 1.4633 s 1.4767 s]
Found 2 outliers among 10 measurements (20.00%)
  1 (10.00%) high mild
  1 (10.00%) high severe

Poseidon2 over (Bn254::Fr, t=3)/1k iter
                        time:   [14.828 ms 14.891 ms 15.010 ms]
Found 2 outliers among 10 measurements (20.00%)
  1 (10.00%) high mild
  1 (10.00%) high severe

Poseidon2 over (Bn254::Fr, t=3)/100k iter
                        time:   [1.4809 s 1.4891 s 1.5012 s]
Found 1 outliers among 10 measurements (10.00%)
  1 (10.00%) high severe
```

File: poseidon2/src/lib.rs (hunk 1 above)
Comment (alxiong) [reply poseidon2/src/lib.rs:7]: done in [6f1678a](https://github.com/EspressoSystems/jellyfish/pull/708/commits/6f1678a8afd66248c625168e07e0595472f71c6b)

**Review by mrain — APPROVED** [review]

**Pull request merged**